File: gof-design-patterns/srcs/P00_04_Abstraction/Car.java
package P00_04_Abstraction;

// 抽象クラス
abstract class Vehicle {
    protected String brand;
    
    public Vehicle(String brand) {
        this.brand = brand;
    }
    
    // 抽象メソッド（実装はサブクラスで定義）
    public abstract void start();
    public abstract void stop();
    
    // 具象メソッド
    public void displayBrand() {
        System.out.println("ブランド: " + brand);
    }
}

// 具象クラス
public class Car extends Vehicle {
    public Car(String brand) {
        super(brand);
    }
    
    @Override
    public void start() {
        System.out.println(brand + "の車がエンジンを始動しました");
    }
    
    @Override
    public void stop() {
        System.out.println(brand + "の車が停止しました");
    }


    public static void main(String[] args) {
        Car car = new Car("Toyota");
        car.start();
        car.stop();
        car.displayBrand();
    }
}

File: gof-design-patterns/typescripts/P00_04_Abstraction/Car.ts
// 抽象クラス
abstract class Vehicle {
    protected brand: string;  // protectedな変数
    
    constructor(brand: string) {
        this.brand = brand;
    }
    
    // 抽象メソッド（実装はサブクラスで定義）
    abstract start(): void;
    abstract stop(): void;
    
    // 具象メソッド
    displayBrand(): void {
        console.log(`ブランド: ${this.brand}`);
    }
}

// 具象クラス
class Car extends Vehicle {
    constructor(brand: string) {
        super(brand);  // 親クラスのコンストラクタを呼び出す
    }
    
    start(): void {
        // 抽象メソッドの実装
        console.log(`${this.brand}の車がエンジンを始動しました`);
    }
    
    stop(): void {
        // 抽象メソッドの実装
        console.log(`${this.brand}の車が停止しました`);
    }
}

// 実行例
const car = new Car("Toyota");
car.start();
car.stop();
car.displayBrand();
